src/solver/marginrankingloss/forward_marginrankingloss.cpp

```diff
@@ -0,0 +1,308 @@
+        {
+            size_t xlocalsize = LOCAL_SIZE_REDUCE;
+            size_t xgridsize  = AlignUp(_size, xlocalsize);
+            size_t ylocalsize = 1;
+            size_t ygridsize  = 1;
+            size_t zlocalsize = 1;
+            size_t zgridsize  = 1;
+
+            auto kernel        = KernelInfo{};
+            kernel.kernel_file = "MIOpenReduceSum.cpp";
+            kernel.kernel_name = "ReduceSumFLOATACCUM";
+
+            kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+            kernel.l_wk.push_back(xlocalsize);
+            kernel.l_wk.push_back(ylocalsize);
+            kernel.l_wk.push_back(zlocalsize);
+
+            kernel.g_wk.push_back(xgridsize);
+            kernel.g_wk.push_back(ygridsize);
+            kernel.g_wk.push_back(zgridsize);
+
+            result.construction_params.push_back(kernel);
+            _size = (_size + LOCAL_SIZE_REDUCE - 1) / LOCAL_SIZE_REDUCE;
+        }
+
+        // Last kernel reduce: FLOAT_ACCUM -> FLOAT
+        size_t xlocalsize = LOCAL_SIZE_REDUCE;
+        size_t xgridsize  = AlignUp(_size, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        auto kernel        = KernelInfo{};
+        kernel.kernel_file = "MIOpenReduceSum.cpp";
+        kernel.kernel_name = "ReduceSum";
+
+        kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(kernel);
+    }
+    // End building result.construction_params
+
+    // Start building result.invoker_factory
+    float divisor = 1.0f;
+    if(problem.GetReductionMode() == MIOPEN_MARGINRANKINGLOSS_REDUCTION_NONE)
+    {
+        // Reduction = None -> invoke 1 kernel
+        result.invoker_factory = [=](const std::vector<Kernel>& kernels) {
+            return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+                decltype(auto) kernel = handle_.Run(kernels.front());
+                decltype(auto) params =
+                    raw_params.CastTo<miopen::marginrankingloss::FwdInvokeParams>();
+
+                auto input1_tv = get_inner_expanded_tv<5>(deref(params.input1Desc));
+                auto input2_tv = get_inner_expanded_tv<5>(deref(params.input2Desc));
+                auto target_tv = get_inner_expanded_tv<5>(deref(params.targetDesc));
+                auto output_tv = get_inner_expanded_tv<5>(deref(params.outputDesc));
+
+                kernel(params.input1,
+                       params.input2,
+                       params.target,
+                       params.output,
+                       params.margin,
+                       divisor,
```

Comment:
`divisor` doesn't need to be a parameter. You can determine it value inside kernel, right?

Reply:
same